### PR TITLE
Pull Request Testing-with notes added in setting up and pre-processing

### PR DESCRIPTION
**Pull request by susiefeng**
Testing;
Only code cell modified is subset parameter, created the dictionary at the parameter setting step.

### CODE REVIEW

File: pipeline_noted.ipynb [hunk 1]
```diff
@@ -14,45 +14,56 @@
...
     "-  an inline `code` usually refer to a name that already exsist in the [namespace](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces) (i.e. the context where we run the codes in this document). It can be a previously encountered concepts, but more often it referes to variable or method names that we [imported](https://docs.python.org/3/reference/import.html) or defined along the way.\n",
-    "-  **bold** texts are used more loosely to highlight anything that requires more attention than plain texts. Though it's not used as carefully as previous formats, it often refer to some specific values that a variable or method arguments can assume."
+    "-  **bold** texts are used more loosely to highlight anything that requires more attention than plain texts, for example, the direct insructions for users. Though it's not used as carefully as previous formats, it often refer to some specific values that a variable or method arguments can assume."
```
Comment (phildong) [inline pipeline_noted.ipynb]: that wasn't the original plan though. Maybe it's time for some "Tips" box?

File: pipeline_noted.ipynb [hunk 2]
```diff
@@ -14,45 +14,56 @@
...
+    "\n",
+    "![workflow](Workflow.PNG)"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "The cells under this section should be executed every time the kernel is restarted."
+    "# <font size=\"9\">**Setting up**</font>\n"
```
Comment (phildong) [inline pipeline_noted.ipynb]: this doesn't seem to show up correctly at least on github. Can you double check on jupyter?

File: pipeline_noted.ipynb [hunk 3]
```diff
@@ -99,14 +110,18 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Here we set all the parameters that controls how the notebook will behave. Ideally, the following cell would be the only thing you have to change when analyzing different datasets. We put all parameters here under a single cell to facilitate batch processing, but most of them will not make sense until we reach and discuss the corresponding steps. So here we discuss only some initial basic parameters that has broader impact, and leave the discussion of specific parameters to each step.\n",
+    "Here we set **all the parameters** that controls how the notebook will behave. Ideally, the following cell would be the only thing you have to change when analyzing different datasets. We put all parameters here under a single cell to facilitate batch processing, but most of them will not make sense until we reach and discuss the corresponding steps. So in this subsecsion, we discuss only some **initial basic parameters** that has broader impact, and leave the discussion of specific parameters to each step.\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: why bold it? That's not a well-defined term of any sort

File: pipeline_noted.ipynb [hunk 4]
```diff
@@ -99,14 +110,18 @@
...
-    "* Both `subset` and `subset_mc` are used to subset the data. `subset` operates on the whole data. A good usecase for this is to take out some troublesome frames or take out some bad pixels, for example, those used as timestamps in wireless miniscope recording. `subset_mc` operates only on the input data to motion correction, and thus only affect how the shifts are calculated. A good usecase for this is if you have good landmark (e.g blood vessels) in the field of view and want to `estimate_shifts` only on those region of interest. Note that if you are running in interactive mode (`interactive = True`), there is an interactive viewer later to help you find proper `subset_mc`. Both `subset` and `subset_mc` should be specified as python [dictionary](https://docs.python.org/3/tutorial/datastructures.html#dictionaries), whose `keys` are the dimension along which subsetting should be done, and `values` specify how subsetting should be done (usually a [slice object](https://docs.python.org/3/c-api/slice.html)). If the word \"dimension\" and how data are indexed are confusing, move on to the section where the data structure are discussed.\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: why is this taken out?

File: pipeline_noted.ipynb [hunk 5]
```diff
@@ -99,14 +110,18 @@
...
+    "* `output_size` controls the realtive size of all the plots on a scale of 0-100 percent, though it can goes beyond 100 without any problem. Adjust this to please your eye.\n",
+    "* `param_save_minian` specify the data saving property for all the saving part. `dpath` is where you want the data to be saved, we recommand use the same dpath as where you load the imaging data. `fname` is the name of your dataset. In `backend`, `'zarr'` is designed for parallel and out-of-core computation, but its support is experimental for now and does not support incremental writing (i.e. it will be a pain to update part of exsisting dataset). `meta_dict` is a `dictionary` that is used to construct meta data for the final labeled data structure. Currently we assume data are stored in some structured folders as shown below:\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: the default works with this structure, doesn't mean we have to assume this specific structure

File: pipeline_noted.ipynb [hunk 6]
```diff
@@ -99,14 +110,18 @@
...
     "* `in_memory` specify whether to load the whole dataset into memory.\n",
-    "* `interactive` controls whether interactive plots should be shown. Not that plotting the results implies actual computation to be carried out, and thus could be very inefficient when the data are not in the memory. In real situation you would probably want `in_memory = True` as well if your data can fit in the memory.\n",
-    "* `output_size` controls the realtive size of all the plots on a scale of 0-100 percent, though it can goes beyond 100 without any problem. Adjust this to please your eye."
+    "* `interactive` controls whether interactive plots should be shown. In real situation, when you want to visualize interactive plots while figuring out the best parameters, you would want `in_memory=True` and `interactive=True` as long as your data can fit in the memory. On the other hand, once you finalized your parameters and ready for batch processing, you would want both of them to be set as False.\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: I still want the user to understand why those two parameters usually go together and what happens if they don't

File: pipeline_noted.ipynb [hunk 7]
```diff
@@ -119,20 +134,31 @@
...
+    "    'overwrite': True}\n",
+    "\n",
+    "#Pre-processing Parameters#\n",
     "param_load_videos = {\n",
     "    'pattern': 'msCam[0-9]+\\.avi$',\n",
     "    'dtype': np.float32,\n",
     "    'in_memory': in_memory,\n",
     "    'downsample': dict(frame=2),\n",
     "    'downsample_strategy': 'subset'}\n",
+    "subset = {\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: this assumes too much. It won't work, for example, if the coordinates doesn't start from 0, (like when it's reversed)

File: pipeline_noted.ipynb [hunk 8]
```diff
@@ -220,13 +255,7 @@
     "    'scs_fallback': False,\n",
     "    'post_scal': True}\n",
     "param_second_merge = {\n",
-    "    'thres_corr': 0.9}\n",
-    "param_save_minian = {\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: why is this taken out?

File: pipeline_noted.ipynb [hunk 9]
```diff
@@ -308,7 +337,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# Pre-processing"
+    "# **Pre-processing**"
```
Comment (phildong) [inline pipeline_noted.ipynb]: why some of the titles are bolded while some are not?

File: pipeline_noted.ipynb [hunk 10]
```diff
@@ -358,7 +387,12 @@
...
+    "\n",
+    "`param_load_videos` should be specified as python [dictionary](https://docs.python.org/3/tutorial/datastructures.html#dictionaries), whose `keys` are the dimension along which subsetting should be done, and `values` specify how subsetting should be done. **If you do want to modify the parameters, you can either go back to the initial parameter setting and change it there. Or by recalling the parameter here, then modify it by changing the values in this specific parameter dictionary.** For example, if you want to stop downsampling your data, you can type in the console: \n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: I don't think we have properly introduced consoles, and it's a pain to attach a console to a running kernel in traditional jupyter notebook (not jupyterlab)

File: pipeline_noted.ipynb [hunk 11]
```diff
@@ -585,7 +632,7 @@
...
...ep carry out denoising of the image frame by frame. The method `denoise` takes in two required arguments: the first is the video array `varr`, and the second `method` is a string specifying the denoising method we are gonna use. Right now three methods are supported: `'gaussian'`, `'median'` and `'anisotropic'`. Under the hood, `denoise` simply call another function frame by frame in a parallel fashion. For `method='gaussian'` it calls [GaussianBlur](https://www.docs.opencv.org/3.3.0/d4/d86/group__imgproc__filter.html#gaabe8c836e97159a9193fb0b11ac52cf1) from `OpenCV` package. For`method='median'` it calls [MedianBlur](https://docs.opencv.org/3.4.3/d4/d86/group__imgproc__filter.html#ga564869aa33e58769b4469101aac458f9) from `OpenCV` package. For `method='anisotropic'` it calls [anisotropic_diffusion](http://loli.github.io/medpy/generated/medpy.filter.smoothing.anisotropic_diffusion.html) from `medpy` package. All additional [keyword arguments](https://docs.python.org/3.7/tutorial/controlflow.html#keyword-arguments) passed into `denoise` are directly passed into one of those two denoising functions under the hood. **Most of the cases, users can simply run through this part of code.**"
```
Comment (phildong) [inline pipeline_noted.ipynb]: that's not true, median filters can give unexpected results some time. I'm working on making the pre-processing part more interactive

File: pipeline.ipynb [hunk 12]
```diff
@@ -64,20 +64,31 @@
...
    "outputs": [],
    "source": [
+    "#Set up Initial Basic Parameters#\n",
     "minian_path = \".\"\n",
     "dpath = \"./demo_movies\"\n",
     "chunks = {\"frame\": 1000, \"height\": 50, \"width\": 50, \"unit_id\": 100}\n",
-    "subset = None\n",
-    "subset_mc = None\n",
-    "in_memory = False\n",
+    "in_memory = True\n",
```
Comment (phildong) [inline pipeline.ipynb]: why is in_memory turned to True?

File: pipeline_noted.ipynb [hunk 13]
```diff
@@ -14,45 +14,56 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "# About this document\n",
+    "<font size=\"6\">**About this document**</font>\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: why do you want to specify the formatting here instead of letting jupyter decide?

File: pipeline_noted.ipynb [hunk 14]
```diff
@@ -14,45 +14,56 @@
...
     "\n",
-    "-  a [hyperlink](https://en.wikipedia.org/wiki/Hyperlink) usually point to a well-defined python module or class or methods, especially when that concept is first encountered in this document and the reader would likely need some background reference. The link usually points to the official documentation of that concept, which might not be the best place to start for beginner. If you find the documentation puzzling, try to google that concept and find a tutorial that best suit your preference.\n",
+    "-  a [hyperlink](https://en.wikipedia.org/wiki/Hyperlink) usually point to a well-defined python module or class or methods, especially when that concept is first encountered in this document and the reader would likely need some background reference. The link usually poing to the official documentation of that concept, which might not be the best place to start for beginner. If you find the documentation puzzling, try to google that concept and find a tutorial that best suit your preference.\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: typo?

File: pipeline_noted.ipynb (hunk 1 above)
Comment (phildong) [reply pipeline_noted.ipynb]: fyi https://nbsphinx.readthedocs.io/en/0.4.2/markdown-cells.html#Info/Warning-Boxes

File: pipeline_noted.ipynb [hunk 15]
```diff
@@ -444,7 +534,11 @@
...
+    "Before proceeding to pre-processing, it's good practice to check if there is something obviously wrong with the video (like camera suddenly drops dark). This can usually be observed by visualizing the video and check the mean fluorescence plot. To take out some bad `frame`, let's say, `frame` after 800, we can utilize the [xarray.DataArray.sel](http://xarray.pydata.org/en/stable/generated/xarray.DataArray.sel.html) method and [slice](https://docs.python.org/3/library/functions.html#slice):\n",
+    "\n",
+    "<div class=\"alert alert-success\">                                                               \n",
+    "The following cell is a example when you want to get rid of the frames after 800: \n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: might as well just get rid of that code cell and inline the code as part of text instruction

File: pipeline_noted.ipynb [hunk 16]
```diff
@@ -480,18 +576,37 @@
...
+    "<div class=\"alert alert-success\">                                                               \n",
+    "Here is good chance to introduce how to create a dictionary. The basic format of a dictionary will be like\n",
+    "</div>\n",
+    "\n",
+    "```python\n",
+    "dictionary = {'key1': value1, 'key2': value2, ... 'keyN': valueN}\n",
+    "```\n",
+    "\n",
+    "<div class=\"alert alert-success\">                                                               \n",
+    "For example, if you only want to keep the first 800 frames, and the height and width from 100 to 200, you can create the subset dictionary like this:\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: want to introduce slice object as well? (and say it doesn't has to be a slice object -- anything `.sel` can take would be fine?)

File: pipeline_noted.ipynb [hunk 17]
```diff
@@ -507,7 +622,10 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Next we drop along `frame` dimension `where` the frame is absolutely dark (the `sum` of all pixel values are zero). Note this step cannot be lazy, i.e. the actual `sum` computation has to be carried out here, so it might not be very efficient if `in_memory=False`. Feel free to skip this step if you are sure there is no such black frames in your data."
+    "Next we drop along `frame` dimension `where` the frame is absolutely dark (the `sum` of all pixel values are zero). Note this step cannot be lazy, i.e. the actual `sum` computation has to be carried out here, so it might not be very efficient if `in_memory=False`. \n",
+    "<div class=\"alert alert-success\">                                                               \n",
+    "Feel free to skip this step if you are sure there is no such black frames in your data.\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: I think you're over-doing this... it's gonna feel so out-of-place with a box containing single sentence that flows well with the context

File: pipeline_noted.ipynb [hunk 18]
```diff
@@ -99,14 +122,37 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "Here we set all the parameters that controls how the notebook will behave. Ideally, the following cell would be the only thing you have to change when analyzing different datasets. We put all parameters here under a single cell to facilitate batch processing, but most of them will not make sense until we reach and discuss the corresponding steps. So here we discuss only some initial basic parameters that has broader impact, and leave the discussion of specific parameters to each step.\n",
+    "Here we set **all the parameters** that controls how the notebook will behave. Ideally, the following cell would be the only thing you have to change when analyzing different datasets. We put all parameters here under a single cell to facilitate batch processing, but most of them will not make sense until we reach and discuss the corresponding steps. So in this subsecsion, we discuss only some initial basic parameters that has broader impact, and leave the discussion of specific parameters to each step.\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: no reason to bold

File: pipeline_noted.ipynb [hunk 19]
```diff
@@ -99,14 +122,37 @@
...
-    "* `output_size` controls the realtive size of all the plots on a scale of 0-100 percent, though it can goes beyond 100 without any problem. Adjust this to please your eye."
+    "* `interactive` controls whether interactive plots should be shown.  Not that plotting the results implies actual computation to be carried out, and thus could be very inefficient when the data are not in the memory. \n",
+    "<div class=\"alert alert-info\">\n",
+    "In real situation, when you want to visualize interactive plots while figuring out the best parameters, you would want in_memory=True and interactive=True as long as your data can fit in the memory. On the other hand, once you finalized your parameters and ready for batch processing, you would want both of them to be set as False.\n",
+    "</div>\n",
+    "\n",
+    "\n",
+    "* `output_size` controls the realtive size of all the plots on a scale of 0-100 percent, though it can goes beyond 100 without any problem.\n",
+    "<div class=\"alert alert-info\">\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: let's take out all "instruction box" with one line, it's too much.

File: pipeline_noted.ipynb [hunk 20]
```diff
@@ -99,14 +122,37 @@
...
+    " Adjust this to please your eye.\n",
+    "</div> \n",
+    "\n",
+    "\n",
+    "* `param_save_minian` specify the data saving property for all the saving part. `dpath` is where you want the data to be saved, we recommand use the same dpath as where you load the imaging data. `fname` is the name of your dataset. In `backend`, `'zarr'` is designed for parallel and out-of-core computation, but its support is experimental for now and does not support incremental writing (i.e. it will be a pain to update part of exsisting dataset). `meta_dict` is a `dictionary` that is used to construct meta data for the final labeled data structure. \n",
+    "<div class=\"alert alert-info\">\n",
+    "The defult meta_dict is assuming data is stored in some structured folders as shown below. We do <strong> recommand </strong> users to structure their data like this to make the following steps (cross-registration) easier. However, you can simply adjust the value of meta_dict in this parameter to fit your data!\n",
```
Comment (phildong) [inline pipeline_noted.ipynb]: cross-registration does not care about folder structure at all

File: pipeline_noted.ipynb [hunk 21]
```diff
@@ -14,45 +14,68 @@
...
     "-  an inline `code` usually refer to a name that already exsist in the [namespace](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces) (i.e. the context where we run the codes in this document). It can be a previously encountered concepts, but more often it referes to variable or method names that we [imported](https://docs.python.org/3/reference/import.html) or defined along the way.\n",
-    "-  **bold** texts are used more loosely to highlight anything that requires more attention than plain texts. Though it's not used as carefully as previous formats, it often refer to some specific values that a variable or method arguments can assume."
+    "-  **bold** texts are used more loosely to highlight anything that requires more attention than plain texts, for example, the direct insructions for users. Though it's not used as carefully as previous formats, it often refer to some specific values that a variable or method arguments can assume.\n",
```
[the remark this answers is not in the document]
Comment (susiefeng) [reply pipeline_noted.ipynb]: ok

File: pipeline_noted.ipynb (hunk 18 above)
Comment (susiefeng) [reply pipeline_noted.ipynb]: ok

File: pipeline_noted.ipynb (hunk 19 above)
Comment (susiefeng) [reply pipeline_noted.ipynb]: well, we can  talk about this

File: pipeline_noted.ipynb (hunk 20 above)
Comment (susiefeng) [reply pipeline_noted.ipynb]: It does, on the level that you need to keep meta_dict consistent

File: pipeline.ipynb [hunk 22]
```diff
@@ -64,20 +64,31 @@
...
    "source": [
+    "#Set up Initial Basic Parameters#\n",
     "minian_path = \".\"\n",
-    "dpath = \"./demo_movies\"\n",
+    "dpath = \"./demo_movies/\"\n",
     "chunks = {\"frame\": 1000, \"height\": 50, \"width\": 50, \"unit_id\": 100}\n",
-    "subset = None\n",
-    "subset_mc = None\n",
-    "in_memory = False\n",
+    "in_memory = True\n",
```
Comment (phildong) [inline pipeline.ipynb]: False

**Comment by phildong** [comment]
closing in favor of a9e5c37a712b1aba801eaf0441838ecb421153f4. Thank you for your work @susiefeng! Your spirit lives safely in the history of `master` now

**Pull request closed without merging**